#----------------------------------------------------------------------
# Run PRScs analysis on the ABCD dataset, using GenLang GWAS sumstats for reading as base datasets
#----------------------------------------------------------------------
# PRScs:
# PRS-CS writes posterior SNP effect size estimates for each chromosome to the user-specified directory. 
# The output file contains chromosome, rs ID, base position, A1, A2 and posterior effect size estimate for each SNP. 
#----------------------------------------------------------------------
# run per chromosome --> array
## example run:
# qsub -v ref_dir=${ref_dir},bim_dir=${project_dir}/prsice/${data}.qc,base_file=${base_dir}${base_pheno}_ed.txt,N=${N},out_dir=${out_dir},target_data=${data} PRScs_run_array.sh
module load plink/6.15
ref_dir=/export/home/acarrion/acarrion/projects/resources/reference_data/1KG/phase3/LD_ref/ldblk_1kg_eur
#----------------------------------------------------------------------
# define variables
root=ABCD
batch=release_3.0_QCed
data=${root}_${batch}
base_project=sibGWAS_Howe2022
#----------------------------------------------------------------------
# build parameters
#----------------------------------------------------------------------
add_dir=/resources/datasets/
analysis_dir=/export/home/acarrion/acarrion/projects/general_scripts/genotyping/PRS/
project_dir=/export/home/acarrion/acarrion/projects/${add_dir}${root}/data/working_data/genotyping/
target_dir=${project_dir}/imputed_genotypes/clean/
#
out_dir=${project_dir}/PRScs/${base_project}/
mkdir -p ${out_dir}
#----------------------------------------------------------------------
# QC of target files
if [ ! -f ${project_dir}/prsice/${data}.qc.bim ]
then
 plink --bfile ${target_dir}${data}.nodup.r2_rs_hg19 \
    --maf 0.05 \
    --mind 0.1 \
    --geno 0.1 \
    --hwe 1e-6 \
    --make-just-bim \
    --make-just-fam \
    --out ${project_dir}/prsice/${data}.qc
fi
#----------------------------------------------------------------------
# Clean base datasets
base_dir=/export/home/acarrion/acarrion/projects/resources/datasets/GWAS_sumstats/QC/ieu-openGWAS/ # *hemen nago*
cd ${base_dir}
base_phenos2test=$(ls *ieu-b-*QC.txt.gz | sed 's/.QC.txt.gz//g') # WR_Z_EUR_combined_STERR_GCOFF_noABCD
# copy and edit base gwas file - one by one, because the input files are not the same
for base_pheno in ${base_phenos2test}
 do
 # define base file
  base_file=${base_dir}/${base_pheno}.QC.txt.gz
  zcat ${base_file} > ${base_pheno}.txt
  sed -i -e 's/ /\t/g' ${base_pheno}.txt # to ensure that all cols are separated with tabs
  # filter based on N -> define minimum N in which the SNP should be present to be included (for GenLang word reading, where Neur~27,000 - ABCD (Neur ~5000) --> 20,000
  awk '{print $2,$4,$5,$6,$9}' ${base_pheno}.txt | sed 's/rsid/SNP/g' | sed 's/a_1/A1/g' | sed 's/a_2/A2/g' | sed 's/beta/BETA/g' |
   sed 's/a/A/g' | sed 's/t/T/g' | sed 's/c/T/g' | sed 's/g/G/g' > ${base_pheno}_ed.txt
  rm ${base_pheno}.txt
  # get the sample size for this study
  grep sample_size ${base_dir}/../../downloaded_data/ieu-openGWAS/${base_pheno}_report.html -C 1 | \
   tail -n 1 | awk -F">" '{print $2}'  | awk -F"<" '{print $1}' > ${base_pheno}_N.txt
 done
#----------------------------------------------------------------------


#----------------------------------------------------------------------
# Run PRScs
#----------------------------------------------------------------------
cd ${out_dir}
for base_pheno in ${base_phenos2test}
 do
  N=$(cat ${base_dir}${base_pheno}_N.txt)
 # Run PRS-CS per chromosome
 qsub -v ref_dir=${ref_dir},bim_dir=${project_dir}/prsice/${data}.qc,base_pheno=${base_pheno},base_file=${base_dir}${base_pheno}_ed.txt,N=${N},target_data=${data},out_dir=${out_dir}\
  ${analysis_dir}PRScs_run_array.sh
 done
#----------------------------------------------------------------------

#----------------------------------------------------------------------
# Generate individual-level polygenic score - for each base dataset
#----------------------------------------------------------------------
# An individual-level polygenic score can be produced by concatenating output files from all chromosomes 
# and then using PLINK's --score command (https://www.cog-genomics.org/plink/1.9/score). 
# If polygenic scores are generated by chromosome, use the 'sum' modifier so that they can be combined into a genome-wide score.
#----------------------------------------------------------------------
for base_pheno in ${base_phenos2test}
 do
 if [ ! -f ${out_dir}/${data}_${base_pheno}.profile ]
 then
  cat ${out_dir}/${data}_${base_pheno}_pst_eff_a1_b0.5_phi1e-02_chr*.txt > ${out_dir}/${data}_${base_pheno}_pst_eff_a1_b0.5_phi1e-02_chrALL.txt
  plink --bfile ${target_dir}${data}.nodup.r2_rs_hg19 \
   --score ${out_dir}/${data}_${base_pheno}_pst_eff_a1_b0.5_phi1e-02_chrALL.txt 2 4 6 header sum \
   --out ${out_dir}/${data}_${base_pheno}
 fi
 done
#----------------------------------------------------------------------

#----------------------------------------------------------------------
# PARSE output
#----------------------------------------------------------------------
analysis_dir=/export/home/acarrion/acarrion/projects/${add_dir}${root}/scripts/genetics/PRS/PRScs/
analysis_dir2=/export/home/acarrion/acarrion/projects/${add_dir}${root}/scripts/genetics/${base_project}/

Rscript ${analysis_dir}PGScs_QC.R ${analysis_dir2}/base_${base_project}.config --verbose
Rscript ${analysis_dir}PGScs_lmm.R ${analysis_dir2}/base_${base_project}.config FALSE --verbose

# visualize
Rscript ${analysis_dir2}PGScs_${base_project}_viz.R ${analysis_dir2}/base_${base_project}.config


#----------------------------------------------------------------------
# remove intermediate files
cd ${base_dir}/
rm *_ed.txt
# rm *_N.txt
#----------------------------------------------------------------------
